# Hands-on example

In [2]:
pip install mlflow

     |████████████████████████████████| 18.1 MB 3.9 MB/s eta 0:00:01
     |████████████████████████████████| 184 kB 5.3 MB/s eta 0:00:01
     |████████████████████████████████| 224 kB 7.1 MB/s eta 0:00:01
  Using cached Jinja2-3.1.2-py3-none-any.whl (133 kB)
     |████████████████████████████████| 41 kB 78 kB/s s eta 0:00:01
     |████████████████████████████████| 83 kB 669 kB/s eta 0:00:011
     |████████████████████████████████| 2.8 MB 10.1 MB/s eta 0:00:01
     |████████████████████████████████| 93 kB 1.0 MB/s eta 0:00:011
     |████████████████████████████████| 39.0 MB 2.2 MB/s eta 0:00:01    |██▉                             | 3.4 MB 9.5 MB/s eta 0:00:04
     |████████████████████████████████| 148 kB 3.2 MB/s eta 0:00:01
     |████████████████████████████████| 79 kB 3.7 MB/s eta 0:00:011
     |████████████████████████████████| 304 kB 3.3 MB/s eta 0:00:01
     |████████████████████████████████| 96 kB 1.8 MB/s eta 0:00:011
     |████████████████████████████████| 62 kB 673 kB/s eta 0:

In [50]:
import os
import warnings
import sys

import pandas as pd
import numpy as np

from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import ElasticNet

import mlflow
import mlflow.sklearn

## 0. The data

* The data set used in this example is from http://archive.ics.uci.edu/ml/datasets/Wine+Quality
* P. Cortez, A. Cerdeira, F. Almeida, T. Matos and J. Reis.
* Modeling wine preferences by data mining from physicochemical properties. In Decision Support Systems, Elsevier, 47(4):547-553, 2009.

In [30]:
data_path = "data/wine-quality.csv"
data=pd.read_csv(data_path,delimiter=';')
data.sample(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
538            12.9              0.35         0.49             5.8      0.066   
1205            7.2              0.36         0.46             2.1      0.074   
205            12.8              0.30         0.74             2.6      0.095   
24              6.9              0.40         0.14             2.4      0.085   
661             7.5              0.42         0.31             1.6      0.080   
645             7.8              0.64         0.10             6.0      0.115   
1503            6.2              0.44         0.39             2.5      0.077   
2               7.8              0.76         0.04             2.3      0.092   
1345            8.2              0.44         0.24             2.3      0.063   
1411            6.4              0.47         0.40             2.4      0.071   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
538                   5.0                  35.0  1.00140  3.20       0.66   
1205                 24.0                  44.0  0.99534  3.40       0.85   
205                   9.0                  28.0  0.99940  3.20       0.77   
24                   21.0                  40.0  0.99680  3.43       0.63   
661                  15.0                  42.0  0.99780  3.31       0.64   
645                   5.0                  11.0  0.99840  3.37       0.69   
1503                  6.0                  14.0  0.99555  3.51       0.69   
2                    15.0                  54.0  0.99700  3.26       0.65   
1345                 10.0                  28.0  0.99613  3.25       0.53   
1411                  8.0                  19.0  0.99630  3.56       0.73   

      alcohol  quality  
538      12.0        7  
1205     11.0        7  
205      10.8        7  
24        9.7        6  
661       9.0        5  
645      10.1        7  
1503     11.0        6  
2         9.8        5  
1345     10.2        6  
1411     10.6        6

## 1. Tracking experiments

### Tracking stores
MLflow supports two types of backend stores: *file store* and *database-backed* store.

- Local file path (specified as file:/my/local/dir), where data is just directly stored locally. Defaults to `mlruns/`
- Database encoded as <dialect>+<driver>://<username>:<password>@<host>:<port>/<database>. Mlflow supports the dialects mysql, mssql, sqlite, and postgresql. For more details, see SQLAlchemy database uri.
- HTTP server (specified as https://my-server:5000), which is a server hosting an MLFlow tracking server.
- Databricks workspace (specified as databricks or as databricks://<profileName>, a Databricks CLI profile.
    
### Artifact stores
- Amazon S3
- Azure Blob Storage
- Google Cloud Storage
- FTP server
- SFTP Server
- NFS
- HDFS

Start the MLflow tracking server by

```
mlflow server \
    --backend-store-uri /mnt/persistent-disk \
    --default-artifact-root s3://my-mlflow-bucket/ \
    --host 0.0.0.0
    --port 5000
```

or use the default storage method to write to `mlruns/`.

In [31]:
# mlflow server --backend-store-uri mlruns/ --default-artifact-root mlruns/ --host 0.0.0.0 --port 5000
remote_server_uri = "http://localhost:5000" # set to your server URI
mlflow.set_tracking_uri(remote_server_uri)  # or set the MLFLOW_TRACKING_URI in the env

In [32]:
mlflow.tracking.get_tracking_uri()

'http://localhost:5000'

In [33]:
exp_name = "ElasticNet_wine"
mlflow.set_experiment(exp_name)

<Experiment: artifact_location='/home/shanmukha/ml/mlruns/560772809476771473', creation_time=1688996319644, experiment_id='560772809476771473', last_update_time=1688996319644, lifecycle_stage='active', name='ElasticNet_wine', tags={}>

### What do we track?

- **Code Version**: Git commit hash used for the run (if it was run from an MLflow Project)
- **Start & End Time**: Start and end time of the run
- **Source**: what code run?
- **Parameters**: Key-value input parameters.
- **Metrics**: Key-value metrics, where the value is numeric (can be updated over the run)
- **Artifacts**: Output files in any format.

In [51]:
def eval_metrics(actual, pred):
    # compute relevant metrics
    rmse = np.sqrt(mean_squared_error(actual, pred))
    mae = mean_absolute_error(actual, pred)
    r2 = r2_score(actual, pred)
    return rmse, mae, r2

In [52]:



def load_data(data_path):
    data=pd.read_csv(data_path,delimiter=';')

    # Split the data into training and test sets. (0.75, 0.25) split.
    train, test = train_test_split(data)

    # The predicted column is "quality" which is a scalar from [3, 9]
    train_x = train.drop(["quality"], axis=1)
    test_x = test.drop(["quality"], axis=1)
    train_y = train[["quality"]]
    test_y = test[["quality"]]
    return train_x, train_y, test_x, test_y




In [53]:
def train(alpha=0.5, l1_ratio=0.5):
    # train a model with given parameters
    warnings.filterwarnings("ignore")
    np.random.seed(40)

    # Read the wine-quality csv file (make sure you're running this from the root of MLflow!)
    data_path = "data/wine-quality.csv"
    train_x, train_y, test_x, test_y = load_data(data_path)

    # Useful for multiple runs (only doing one run in this sample notebook)
    with mlflow.start_run():
        # Execute ElasticNet
        lr = ElasticNet(alpha=alpha, l1_ratio=l1_ratio, random_state=42)
        lr.fit(train_x, train_y)

        # Evaluate Metrics
        predicted_qualities = lr.predict(test_x)
        (rmse, mae, r2) = eval_metrics(test_y, predicted_qualities)

        # Print out metrics
        print("Elasticnet model (alpha=%f, l1_ratio=%f):" % (alpha, l1_ratio))
        print("  RMSE: %s" % rmse)
        print("  MAE: %s" % mae)
        print("  R2: %s" % r2)

        # Log parameter, metrics, and model to MLflow
        mlflow.log_param(key="alpha", value=alpha)
        mlflow.log_param(key="l1_ratio", value=l1_ratio)
        mlflow.log_metric(key="rmse", value=rmse)
        mlflow.log_metrics({"mae": mae, "r2": r2})
        mlflow.log_artifact(data_path)
        print("Save to: {}".format(mlflow.get_artifact_uri()))

        mlflow.sklearn.log_model(lr, "model")

In [43]:
# could also do
# with mlflow.start_run():
#     for epoch in range(0, 3):
#          mlflow.log_metric(key="quality", value=2*epoch, step=epoch)

In [55]:
train(0.5, 0.5)

Elasticnet model (alpha=0.500000, l1_ratio=0.500000):
  RMSE: 0.793164022927685
  MAE: 0.6271946374319586
  R2: 0.10862644997792636
Save to: /home/shanmukha/ml/mlruns/560772809476771473/b46c2ca44a684ccabd5b97a7c56eae4e/artifacts


In [56]:
train(0.2, 0.2)

Elasticnet model (alpha=0.200000, l1_ratio=0.200000):
  RMSE: 0.7336400911821402
  MAE: 0.5643841279275427
  R2: 0.2373946606358417
Save to: /home/shanmukha/ml/mlruns/560772809476771473/2528eab0e9f14fa18df22e1280e4ebe8/artifacts


In [57]:
train(0.1, 0.1)

Elasticnet model (alpha=0.100000, l1_ratio=0.100000):
  RMSE: 0.7128829045893679
  MAE: 0.5462202174984664
  R2: 0.2799376066653345
Save to: /home/shanmukha/ml/mlruns/560772809476771473/7c6a81778bde434e9e8008f5b19356c9/artifacts


### 1.1 Comparing runs
Run `mlflow ui` in a terminal or `http://your-tracking-server-host:5000` to view the experiment log and visualize and compare different runs and experiments. The logs and the model artifacts are saved in the `mlruns` directory (or where you specified).

## 2. Packaging the experiment as a MLflow project as conda env

Specify the entrypoint for this project by creating a `MLproject` file and adding an conda environment with a `conda.yaml`. You can copy the yaml file from the experiment logs.

To run this project, invoke `mlflow run . -P alpha=0.42`. After running this command, MLflow runs your training code in a new Conda environment with the dependencies specified in `conda.yaml`.


## 3. Deploy the model

Deploy the model locally by running

`mlflow models serve -m mlruns/0/f5f7c052ddc5469a852aa52c14cabdf1/artifacts/model/ -h 0.0.0.0 -p 1234`

Test the endpoint:

`curl -X POST -H "Content-Type:application/json; format=pandas-split" --data '{"columns":["alcohol", "chlorides", "citric acid", "density", "fixed acidity", "free sulfur dioxide", "pH", "residual sugar", "sulphates", "total sulfur dioxide", "volatile acidity"],"data":[[12.8, 0.029, 0.48, 0.98, 6.2, 29, 3.33, 1.2, 0.39, 75, 0.66]]}' http://0.0.0.0:1234/invocations`

You can also simply build a docker image from your model

`mlflow models build-docker -m mlruns/1/d671f37a9c7f478989e67eb4ff4d1dac/artifacts/model/ -n elastic_net_wine`

and run the container with

`docker run -p 8080:8080 elastic_net_wine`.

Or you can directly deploy to AWS sagemaker or Microsoft Azure ML.

## 4. Tagging runs

In [59]:
from datetime import datetime
from mlflow.tracking import MlflowClient

client = MlflowClient()
print(client.get_experiment(560772809476771473))

<Experiment: artifact_location='/home/shanmukha/ml/mlruns/560772809476771473', creation_time=1688996319644, experiment_id='560772809476771473', last_update_time=1688996319644, lifecycle_stage='active', name='ElasticNet_wine', tags={}>


In [61]:
# get the run
_run = client.get_run(run_id="7c6a81778bde434e9e8008f5b19356c9")
print(_run)

<Run: data=<RunData: metrics={'mae': 0.5462202174984664,
 'r2': 0.2799376066653345,
 'rmse': 0.7128829045893679}, params={'alpha': '0.1', 'l1_ratio': '0.1'}, tags={'mlflow.log-model.history': '[{"run_id": "7c6a81778bde434e9e8008f5b19356c9", '
                             '"artifact_path": "model", "utc_time_created": '
                             '"2023-07-10 14:01:39.425233", "flavors": '
                             '{"python_function": {"model_path": "model.pkl", '
                             '"predict_fn": "predict", "loader_module": '
                             '"mlflow.sklearn", "python_version": "3.8.10", '
                             '"env": {"conda": "conda.yaml", "virtualenv": '
                             '"python_env.yaml"}}, "sklearn": '
                             '{"pickled_model": "model.pkl", '
                             '"sklearn_version": "1.0.2", '
                             '"serialization_format": "cloudpickle", "code": '
                             'n

In [62]:
# add a tag to the run
dt = datetime.now().strftime("%d-%m-%Y (%H:%M:%S.%f)")
client.set_tag(_run.info.run_id, "deployed", dt)

# Connect to a postgesql db

Connect to the db:
    
`sudo -u postgres psql`

Create a user and a database for the tracking server:
    
```
CREATE DATABASE mlflow;
CREATE USER mlflow WITH ENCRYPTED PASSWORD 'mlflow';
GRANT ALL PRIVILEGES ON DATABASE mlflow TO mlflow;
```

```
mlflow server --backend-store-uri postgresql://mlflow:mlflow@localhost/mlflow --default-artifact-root file:/home/tobias/Projects/mlflow_talk/pydataberlin-2019/mlruns/ -h 0.0.0.0 -p 8000
```

Run the notebook again with this tracking server.

Look at the db:

`psql mlflow`

`SELECT * FROM experiments;`

`SELECT * FROM runs;`